webapi/src/main/scala/org/knora/webapi/messages/util/search/gravsearch/transformers/OntologyInferencer.scala

```diff
@@ -0,0 +1,176 @@
+import org.knora.webapi.messages.SmartIri
+import org.knora.webapi.messages.StringFormatter
+import org.knora.webapi.messages.util.search._
+import org.knora.webapi.slice.ontology.repo.model.OntologyCacheData
+import org.knora.webapi.slice.ontology.repo.service.OntologyCache
+
+final case class OntologyInferencer(
+  private val ontologyCache: OntologyCache,
+  implicit private val stringFormatter: StringFormatter
+) {
+
+  private def inferSubclasses(
+    statementPattern: StatementPattern,
+    cache: OntologyCacheData,
+    limitInferenceToOntologies: Option[Set[SmartIri]]
+  ): IO[GravsearchException, Seq[QueryPattern]] = for {
+    // Yes. Expand using rdfs:subClassOf*.
+    baseClassIri <-
+      statementPattern.obj match {
+        case iriRef: IriRef => ZIO.succeed(iriRef)
+        case other =>
+          ZIO.fail(GravsearchException(s"The object of rdf:type must be an IRI, but $other was used"))
+      }
+
+    // look up subclasses from ontology cache
+    superClasses = cache.classToSubclassLookup
+    knownSubClasses =
+      superClasses
+        .get(baseClassIri.iri)
+        .getOrElse({
+          Set(baseClassIri.iri)
+        })
+        .toSeq
+
+    // if provided, limit the child classes to those that belong to relevant ontologies
+    subClasses = limitInferenceToOntologies match {
+                   case None                       => knownSubClasses
+                   case Some(relevantOntologyIris) =>
+                     // filter the known subclasses against the relevant ontologies
+                     knownSubClasses.filter { subClass =>
+                       cache.classDefinedInOntology.get(subClass) match {
+                         case Some(ontologyOfSubclass) =>
+                           // return true, if the ontology of the subclass is contained in the set of relevant ontologies; false otherwise
+                           relevantOntologyIris.contains(ontologyOfSubclass)
+                         case None => false // should never happen
+                       }
+                     }
+                 }
+    // if subclasses are available, create a union statement that searches for either the provided triple (`?v a <classIRI>`)
+    // or triples where the object is a subclass of the provided object (`?v a <subClassIRI>`)
+    // i.e. `{?v a <classIRI>} UNION {?v a <subClassIRI>}`
+    x = if (subClasses.length > 1)
+          Seq(UnionPattern(subClasses.map(newObject => Seq(statementPattern.copy(obj = IriRef(newObject))))))
+        else
+          // if no subclasses are available, the initial statement can be used.
+          Seq(statementPattern)
+
+  } yield x
+
+  private def inferSubproperties(
+    statementPattern: StatementPattern,
+    predIri: SmartIri,
+    cache: OntologyCacheData,
+    limitInferenceToOntologies: Option[Set[SmartIri]]
+  ): IO[GravsearchException, Seq[QueryPattern]] = {
+
+    // No. Expand using rdfs:subPropertyOf*.
```

Comment:
???

Reply:
old comment, I thought I had changed it. Done now.